### ISSUE TITLE
Set-up SSL for production

### ISSUE THREAD

**Issue by kevinongko** [codecasts/ambientum#46]
Hi, 

Any guide/tutorial to set up SSL such as `Let's Encrypt`  for production using `ambientum/php:latest-nginx` image?

Thanks :)

**Comment by hernandev** [comment]
### For Development
Ambientum's Nginx images for PHP already comes with full **https** and **http 2** support.

The containers will listen for https connections on port **8083**.

For development purposes, you don't need to do anything else.

---

### For Production

Using either self signed or vendor issued certificates on Ambientum, takes 3 files:

| File Name | Where to Place It | Description |
|--------|-------|-----| 
|`nginx.key`|`/home/ambientum/ssl/nginx.key`|Private Key|
|`nginx.crt`|`/home/ambientum/ssl/nginx.crt`|Certificate|
|`dhparam.pem`| `/home/ambientum/ssl/dhparam.pem`|Diffie-Hellman Param for PFS|

**Those three files exists by default on the images, but you should never that it to production using the sample ones.**

The last file is not so usual but Ambientum Nginx config requires it, learn more about [Diffie-Hellman Key Exchange](https://en.wikipedia.org/wiki/Diffie–Hellman_key_exchange) and [Perfect Forward Secrecy (FPS)](https://en.wikipedia.org/wiki/Forward_secrecy)

> Tip: If you use some build system like travis or gitlab-ci you can easily configure them to generate those files for you at build time.

---- 
### Production Security Tips
**Avoid exposing the container directly** directly, by adopting one or more of the following scenarios:

1) Setup a separate **`Nginx`** or **`HAProxy`** instance that will proxy the requests to the containers, this way, the container is not directly exposed to requests and all valid SSL can be configured there. 
2) Having either **`Cloudflare`** or other CDN handling the https requests.

In both cases, the certificate on the container can be **self-signed** since users will never reach then, only Nginx or HAProxy or Cloudflare.

3) Use both **1** and **2** together **[recommended]** so the flux would be something like: 

```
USER
  >> CLOUDFLARE        [valid]
    >> NGINX/HAPROXY   [self-signed or lets-encript]
      >> CONTAINER     [self-signed]
```

----
### ANNEX

#### Generating a self-signed certificate (3650 days):
```
openssl req -x509 -nodes \
   -days 3650 \
   -newkey rsa:2048 \
   -keyout nginx.key \
   -out nginx.crt \ 
   -subj  "/C=AM/ST=Ambientum/L=Ambientum/O=Ambientum/CN=*.dev"
```

#### Generating a `dhparam.pem` file:
This is needed even if you use a 3rd part certificate like let's encrypt or Comodo

```
openssl dhparam -out dhparam.pem 2048
```

**Comment by kevinongko** [comment]
Hi @hernandev ,

Thanks for the tutorial!

btw I found the `ambientum/php:latest-nginx`'s nginx.conf doesn't have SSL configuration only HTTP.
When I change to `ambientum/php:7.1-nginx` the SSL works.
is this normal?

**Comment by hernandev** [comment]
No, latest-nginx should be pointing to ambientum/php:7.1-nginx.

If you're not getting SSL, please try pulling the image again, it does not updates automatically.

```
docker pull ambientum/php:latest-nginx
```

So you know, `latest-*` are just a convention for any lastest version available, you may wanna set the version explicitly so you don't upgrade without even knowing.

:)

**Comment by kevinongko** [comment]
I try to re-pull again now it's updated with SSL config.

Thanks for the help, I'll close this issue now 👍

**Comment by hernandev** [comment]
I'll reopen this so I remember to include this guide on the docs, don't worry will merge it into a document and close it after that.

Thanks

**Comment by yanccprogramador** [comment]
php:7.3-nginx a pasta para funcionar é /home/ssl não /home/ambientum/ssl
